File: bindings/nim/src/proven/safe_math.nim
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
#
# Safe mathematical operations with overflow detection.
# Thin wrapper over libproven FFI -- all logic lives in Idris.

import std/options
import lib_proven

proc safeDiv*(numerator, denominator: int64): Option[int64] =
  ## Safely divide two integers, returning None on division by zero.
  let res = provenMathDiv(numerator, denominator)
  if res.status == PROVEN_OK:
    return some(res.value)
  none(int64)

proc safeMod*(numerator, denominator: int64): Option[int64] =
  ## Safely compute modulo, returning None on division by zero.
  let res = provenMathMod(numerator, denominator)
  if res.status == PROVEN_OK:
    return some(res.value)
  none(int64)

proc safeAdd*(a, b: int64): Option[int64] =
  ## Safely add two integers with overflow detection.
  ## Returns None if overflow would occur.
  let res = provenMathAddChecked(a, b)
  if res.status == PROVEN_OK:
    return some(res.value)
  none(int64)

proc safeSub*(a, b: int64): Option[int64] =
  ## Safely subtract two integers with overflow detection.
  ## Returns None if overflow would occur.
  let res = provenMathSubChecked(a, b)
  if res.status == PROVEN_OK:
    return some(res.value)
  none(int64)

proc safeMul*(a, b: int64): Option[int64] =
  ## Safely multiply two integers with overflow detection.
  ## Returns None if overflow would occur.
  let res = provenMathMulChecked(a, b)
  if res.status == PROVEN_OK:
    return some(res.value)
  none(int64)

proc safeAbs*(n: int64): Option[int64] =
  ## Safely compute absolute value (handles int64.low).
  ## Returns None if result would overflow.
  let res = provenMathAbsSafe(n)
  if res.status == PROVEN_OK:
    return some(res.value)
  none(int64)

proc clamp*(lo, hi, value: int64): int64 =
  ## Clamp a value to the range [lo, hi].
  provenMathClamp(lo, hi, value)

proc safePow*(base: int64, exp: uint32): Option[int64] =
  ## Safely compute integer exponentiation with overflow detection.
  ## Returns None if overflow would occur.
  let res = provenMathPowChecked(base, exp)
  if res.status == PROVEN_OK:
    return some(res.value)
  none(int64)

File: bindings/haskell/src/Proven/SafeMath.hs
{- SPDX-License-Identifier: PMPL-1.0-or-later -}
{- Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk> -}

-- | Safe mathematical operations via libproven FFI.
--
-- All computation is performed by the Idris 2 verified core.
-- This module only marshals arguments and results.
module Proven.SafeMath
  ( safeDiv
  , safeMod
  , safeAdd
  , safeSub
  , safeMul
  , safeAbs
  , safeClamp
  , safePow
  ) where

import Proven.FFI (c_proven_math_div, c_proven_math_mod,
                   c_proven_math_add_checked, c_proven_math_sub_checked,
                   c_proven_math_mul_checked, c_proven_math_abs_safe,
                   c_proven_math_clamp, c_proven_math_pow_checked)
import Proven.Core (intResultToMaybe)

-- | Safely divide two integers. Returns Nothing on division by zero.
-- Delegates to @proven_math_div@ in libproven.
safeDiv :: Int -> Int -> IO (Maybe Int)
safeDiv a b = intResultToMaybe <$>
  c_proven_math_div (fromIntegral a) (fromIntegral b)

-- | Safely compute modulo. Returns Nothing on division by zero.
-- Delegates to @proven_math_mod@ in libproven.
safeMod :: Int -> Int -> IO (Maybe Int)
safeMod a b = intResultToMaybe <$>
  c_proven_math_mod (fromIntegral a) (fromIntegral b)

-- | Safely add two integers with overflow detection.
-- Delegates to @proven_math_add_checked@ in libproven.
safeAdd :: Int -> Int -> IO (Maybe Int)
safeAdd a b = intResultToMaybe <$>
  c_proven_math_add_checked (fromIntegral a) (fromIntegral b)

-- | Safely subtract two integers with underflow detection.
-- Delegates to @proven_math_sub_checked@ in libproven.
safeSub :: Int -> Int -> IO (Maybe Int)
safeSub a b = intResultToMaybe <$>
  c_proven_math_sub_checked (fromIntegral a) (fromIntegral b)

-- | Safely multiply two integers with overflow detection.
-- Delegates to @proven_math_mul_checked@ in libproven.
safeMul :: Int -> Int -> IO (Maybe Int)
safeMul a b = intResultToMaybe <$>
  c_proven_math_mul_checked (fromIntegral a) (fromIntegral b)

-- | Safe absolute value. Handles MIN_INT correctly.
-- Delegates to @proven_math_abs_safe@ in libproven.
safeAbs :: Int -> IO (Maybe Int)
safeAbs n = intResultToMaybe <$>
  c_proven_math_abs_safe (fromIntegral n)

-- | Clamp a value to the range [lo, hi].
-- Delegates to @proven_math_clamp@ in libproven.
safeClamp :: Int -> Int -> Int -> IO Int
safeClamp lo hi val = fromIntegral <$>
  c_proven_math_clamp (fromIntegral lo) (fromIntegral hi) (fromIntegral val)

-- | Safe integer power with overflow checking.
-- Delegates to @proven_math_pow_checked@ in libproven.
safePow :: Int -> Word -> IO (Maybe Int)
safePow base exponent' = intResultToMaybe <$>
  c_proven_math_pow_checked (fromIntegral base) (fromIntegral exponent')
